Sources/Main.cpp

```diff
@@ -1174,51 +1139,101 @@ void __fastcall TX584Form::PasteItemClick(TObject *Sender)
                 CPU.Format(Code[index + i], str);
                 CodeListView->Items->Item[index + i]->SubItems->Strings[1] = str;
                 CodeListView->Items->Item[index + i]->SubItems->Strings[2] = CMClipboard[i];
+                CodeListView->Items->Item[index + i]->Selected = true;
             }
-        SelCount = ClipboardSize;
-        if (index + SelCount > MAX_ADDR)
-            SelCount = MAX_ADDR - index;
+
         CodeListView->Repaint();
         SetModifyFlag(true);
     } else
         PostMessageW(ActiveControl->Handle, WM_PASTE, 0, 0);
 }
 //---------------------------------------------------------------------------
 
+void TX584Form::ClearSelectedItems()
+{
+    TListItem *Item;
+
+    Item = CodeListView->Selected;
+    TItemStates selected = TItemStates() << isSelected;
+    for (; Item; Item = CodeListView->GetNextItem(Item, sdBelow, selected)) {
+        int Index = Item->Index;
+        Code[Index] = NOP;
+        Item->SubItems->Strings[1] = NOP_TEXT;
+        Item->SubItems->Strings[2] = L"";
+    }
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::RemoveSelectedItems()
+{
+    bool NewItems[MAX_ADDR] = {0};
+
+    for (int i = 0; i < MAX_ADDR; i++) {
+        NewItems[i] = true;
+    }
+
+    TListItem *Item;
+    Item = CodeListView->Selected;
+
+    TItemStates selected = TItemStates() << isSelected;
+
+    for (; Item; Item = CodeListView->GetNextItem(Item, sdBelow, selected)) {
+        int Index = Item->Index;
+        NewItems[Index] = false;
+    }
+
+    int NewIndex = 0;
+    int OldIndex = 0;
+
+    for (; OldIndex < MAX_ADDR && NewIndex < MAX_ADDR; OldIndex++) {
+        if (NewItems[OldIndex]) {
+            TListItem *ItemOld, *ItemNew;
+            Code[NewIndex] = Code[OldIndex];
+            ItemOld = CodeListView->Items->Item[OldIndex];
+            ItemNew = CodeListView->Items->Item[NewIndex];
+            ItemNew->SubItems->Strings[1] = ItemOld->SubItems->Strings[1];
+            ItemNew->SubItems->Strings[2] = ItemOld->SubItems->Strings[2];
+            NewIndex++;
+        }
+    }
+
+    for (; NewIndex < MAX_ADDR; NewIndex++) {
+        Code[NewIndex] = NOP;
+        CodeListView->Items->Item[NewIndex]->SubItems->Strings[1] = NOP_TEXT;
+        CodeListView->Items->Item[NewIndex]->SubItems->Strings[2] = L"";
+    }
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::ClearSelection()
+{
+    if (!CodeListView) {
```

Comment:
Can `CodeListView` ever be null?

Reply:
Yes, it can, if this method called BEFORE CodeListView ever initialized

Reply:
I believe `CodeListView` is created in the parent class constructor, even before `TX584Form` is constructed:

```cpp
__fastcall TX584Form::TX584Form(TComponent* Owner)
    : TForm(Owner) // <== initialization happens here
{
}
```

After that you can access `CodeListView` and any other controls. They are guaranteed to be non-null :smiley:

Reply:
First TX584Form created, then on FormCreate CodeListView is initialized, no?

Reply:
I think I should check without 
```c++
if (!CodeListView)
    return;
````

Reply:
> First TX584Form created, then on FormCreate CodeListView is initialized, no?

Nope, this is not how C++ works :stuck_out_tongue:  You can easily debug the creation order by setting breakpoints at `TX584Form` constructor and `FormCreate` handler.

1. First `TX584Form` constructor is called.
2. It immediately calls parent class constructor `TForm` before doing anything else.
3. Parent class constructor loads the form and initializes all controls, including `CodeListView`.
3. Then body of `TX584Form` constructor is executed. At this point `CodeListView` is already initialized. You can inspect its properties in the debugger window.
4. After that `FormCreate` is called.